#166265856 Lend a book

ocranbillions opened the pull request:
### What does this PR do?
- An admin user can be able to lend out books to patrons

#### Description of Task to be completed?
- Write unit tests
- create model for borrowedBooks
- add /borrow endpoint to books route
- add lendbook method as routeHandler to bookController
- get form field from req.body
- check db for books not yet returned by the patron
- check if patron has borrowed a maximum of 3 books
- check if patron has a fine on any of the borrowed books
- proceed with the lending process if patron passes all checks
- change the book status to `borrowed` in the books table
- document api

#### How should this be manually tested?
- Login as an admin and register a book as borrowed by making a post request to /books/borrow with `isbn` and `patronId`  on the request body

#### Any background context you want to provide?
- N/A

#### What are the relevant pivotal tracker stories?
- [#166265856](https://www.pivotaltracker.com/story/show/166265856)

#### Screenshots (if appropriate)
![Screenshot 2019-06-13 at 7 01 00 PM](https://user-images.githubusercontent.com/29557902/59464535-99af2200-8e20-11e9-83ed-e7e202b1fc68.png)
![Screenshot 2019-06-13 at 7 02 53 PM](https://user-images.githubusercontent.com/29557902/59464536-99af2200-8e20-11e9-8e64-bcff3ff57997.png)
![Screenshot 2019-06-13 at 7 03 00 PM](https://user-images.githubusercontent.com/29557902/59464538-99af2200-8e20-11e9-8a6c-fc2be7a6f36c.png)
![Screenshot 2019-06-13 at 7 03 05 PM](https://user-images.githubusercontent.com/29557902/59464539-9a47b880-8e20-11e9-8bc7-f919e6ae892c.png)
![Screenshot 2019-06-13 at 7 07 44 PM](https://user-images.githubusercontent.com/29557902/59464541-9a47b880-8e20-11e9-948a-7afa3246f498.png)



#### Questions:
- N/A

ocranbillions committed:
feature(book): lend book

Write unit tests
create model for borrowedBooks
add /borrow endpoint to books route
add lendbook method as routeHandler to bookController
get form field from req.body
check db for books not yet returned by the patron
check if patron has borrowed a maximum of 3 books
check if patron has a fine on any of the borrowed books
proceed with the lending process if patron passes all checks
document api

[Delivers ##166265856]

ocranbillions committed:
feature(book): lend book

Write unit tests
create model for borrowedBooks
add /borrow endpoint to books route
add lendbook method as routeHandler to bookController
get form field from req.body
check db for books not yet returned by the patron
check if patron has borrowed a maximum of 3 books
check if patron has a fine on any of the borrowed books
proceed with the lending process if patron passes all checks
document api

[Delivers ##166265856]

PelumiAlesh commented on server/controllers/bookController.js:
```diff
@@ -90,6 +90,50 @@ class BookController {
+      const booksNotReturned = await models.Borrowed_books.findAndCountAll({ where: { patronId, returned : false } });
+
+      if (booksNotReturned.count >= 3) {
+        return Utils.errorStatus(res, 405, 'You cannot have more than 3 books in your possession');        
```
This error should be `'You cannot have 3 or more books in your possession' because you are checking if the book is 3 or higher

ocranbillions replied:
The user can have a maximum of 3 books in their possession so the message should suffice.

PelumiAlesh replied:
Yes, But I was referring to the error message itself. It says you cannot have more than 3 books, but you can not have 3 books as well.

chuxmykel replied:
@PelumiAlesh I think it's cool. You can have 3. But not more

PelumiAlesh replied:
@chuxmykel He is using **>=** in ` if (booksNotReturned.count >= 3)` which means you can not have 3.

chuxmykel replied:
How @PelumiAlesh? Isn't that supposed to be greater than or equal 3? Which means 3 or above now?

chuxmykel replied:
It means the books he has not returned are 3 or more. So why do you think he can't have 3?

ocranbillions replied:
@PelumiAlesh by the time that if condition evaluates to true, the user would have had 3 books already. That's when the response is returned.

PelumiAlesh replied:
@chuxmykel explained to me. Yeah its good, sorry for the confusion :)

PelumiAlesh commented on server/controllers/bookController.js:
```diff
@@ -90,6 +90,50 @@ class BookController {
+      if(overdueBook) {
+        const dueDate = overdueBook.dueDate.toString().substring(0, 15);
+        const responseObj = {
+          message: `The book titled ${overdueBook.title.toUpperCase()} was due for return on ${dueDate}. Kindly return before we can proceed with a new request`
```
When the patron has 2 books that are overdue, it only returns error message that the first book is overdue. Find a way to loop through it if the overdue books are 2... If it is three or more the first validation `You cannot have more than 3 books in your possession` would cover that.

ocranbillions replied:
This too.

PelumiAlesh commented on server/controllers/bookController.js:
```diff
@@ -90,6 +90,50 @@ class BookController {
   }
+
+  static async lendBook(req, res) {
+    const { isbn, title, patronId, cost} = req.body;
```
- You should validate if the patron actually exists in the database and also reference the user's table.

- I don't think you need the title because you are only checking with the ISBN. 

- And you should validate the ISBN too if they exist on the database

PelumiAlesh commented on server/controllers/bookController.js:
```diff
@@ -90,6 +90,50 @@ class BookController {
+        dateBorrowed,
+        dueDate,
+        patronId,
+        cost
```
I think you should also remove this cost because it might conflict with the real amount on the book's table.

P.s: Abraham said we should make the platform a monthly payment, Books would no longer be paid for again. All the patron pays for is their monthly bill, damages to book, and overdue books.

ocranbillions replied:
Noted bro, will work on it soon. Thanks

chuxmykel reviewed (approved):
LGTM

chuxmykel reviewed (dismissed):
Lgtm

ocranbillions committed:
feature(book): remove cost from borrowed_books model

chuxmykel reviewed (dismissed):
LGTM

emp-daisy commented on server/controllers/bookController.js:
```diff
@@ -90,6 +90,49 @@ class BookController {
+    const { isbn, title, patronId} = req.body;
+    
+    try {
+      const booksNotReturned = await models.Borrowed_books.findAndCountAll({ where: { patronId, returned : false } });
```
Borrowed_books??? what casing is this?

emp-daisy commented on server/db/models/borrowed_books.js:
```diff
@@ -0,0 +1,32 @@
+/* eslint-disable camelcase */
+export default (sequelize, DataTypes) => {
+  const Borrowed_books = sequelize.define('Borrowed_books', {
```
Borrowed_books???
Stick to camel case. "BorrowedBooks"

ocranbillions replied:
ok

emp-daisy commented on server/docs/hellobooks_api_doc.yaml:
```diff
@@ -543,6 +543,55 @@ paths:
+                  type: string
+                patronId:
+                  type: integer
+                cost:
```
Where is this used?

ocranbillions replied:
I'll remove it, it's not meant to be there. I initially had it in the model but I later removed it because we've concluded that payments will be based on monthly subscription.

emp-daisy commented on server/routes/bookRoute.js:
```diff
@@ -12,4 +12,6 @@ bookRoute.delete('/:id', Validate.id, BookController.deleteBook);
 bookRoute.post('/request', 
 Authenticate.isLoggedIn, Validate.requestBook, BookController.requestBook)
 
+bookRoute.post('/borrow', Authenticate.isLoggedIn, Authenticate.isAdmin, Validate.lendBook, BookController.lendBook);
```
Is this only available to admins? What about super admins?

ocranbillions replied:
I implemented isAdmin to check for both admin and super admin so the route is open to both of these people.

PelumiAlesh committed:
bg(validation): fix validation and naming

add validations to check if isbn exist
add validations to check if patronId exist
fix wrong naming conventions
fix the book title returned
update the swagger doc

PelumiAlesh committed:
bg(validation): fix validation bug

PelumiAlesh committed:
Update bookController.js

PelumiAlesh committed:
Merge branch 'develop' into ft/166265856/lend-a-book

PelumiAlesh committed:
bg(validation): fix validation

PelumiAlesh committed:
bg(test): fix travis failing test

PelumiAlesh committed:
bg(test): fix travis failing test
